cdap-explore/src/main/java/co/cask/cdap/explore/executor/ExploreExecutorHttpHandler.java

```diff
@@ -142,85 +145,131 @@ public void disableStream(HttpRequest request, HttpResponder responder,
+      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, "Error getting spec for dataset " + datasetId);
       return;
     }
 
     try {
-      QueryHandle handle = exploreTableManager.enableDataset(datasetID, datasetSpec);
+      QueryHandle handle = exploreTableManager.enableDataset(datasetId, datasetSpec);
       JsonObject json = new JsonObject();
       json.addProperty("handle", handle.getHandle());
       responder.sendJson(HttpResponseStatus.OK, json);
     } catch (IllegalArgumentException e) {
       responder.sendString(HttpResponseStatus.BAD_REQUEST, e.getMessage());
     } catch (ExploreException e) {
-      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, "Error enabling explore on dataset " + datasetID);
+      responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, "Error enabling explore on dataset " + datasetId);
     } catch (SQLException e) {
       responder.sendString(HttpResponseStatus.BAD_REQUEST,
-                           "SQL exception while trying to enable explore on dataset " + datasetID);
+                           "SQL exception while trying to enable explore on dataset " + datasetId);
     } catch (UnsupportedTypeException e) {
       responder.sendString(HttpResponseStatus.BAD_REQUEST,
-                           "Schema for dataset " + datasetID + " is not supported for exploration: " + e.getMessage());
+                           "Schema for dataset " + datasetId + " is not supported for exploration: " + e.getMessage());
     } catch (Throwable e) {
       LOG.error("Got exception:", e);
       responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR, e.getMessage());
     }
   }
 
   /**
-   * Disable ad-hoc exploration of a dataset instance.
+   * Enable ad-hoc exploration of a dataset instance.
    */
   @POST
-  @Path("datasets/{dataset}/disable")
-  public void disableDataset(HttpRequest request, HttpResponder responder,
-                             @PathParam("namespace-id") String namespaceId, @PathParam("dataset") String datasetName) {
+  @Path("datasets/{dataset}/update")
+  public void updateDataset(HttpRequest request, HttpResponder responder,
+                            @PathParam("namespace-id") String namespace, @PathParam("dataset") String datasetName)
+    throws BadRequestException {
 
-    LOG.debug("Disabling explore for dataset instance {}", datasetName);
-    Id.DatasetInstance datasetID = Id.DatasetInstance.from(namespaceId, datasetName);
-    DatasetSpecification spec;
+    DatasetId datasetId = new DatasetId(namespace, datasetName);
+    DatasetSpecification oldSpec = readDatasetSpec(request);
+    DatasetSpecification datasetSpec;
     try {
-      spec = datasetFramework.getDatasetSpec(datasetID);
-      if (spec == null) {
-        responder.sendString(HttpResponseStatus.NOT_FOUND, "Cannot load dataset " + datasetID);
+      datasetSpec = datasetFramework.getDatasetSpec(datasetId.toId());
```

Comment:
For enable, we probably want to keep the endpoint that enables explore without passing in a spec, because that is usable through the router to enable explore for a dataset outside the context of creating it. I can add an endpoint /enable-internal that expects the spec in the body.